File: trees/binarySearchTree.js
class Node{
    constructor(val){
        this.val = val;
        this.right = null;
        this.left = null;
    }
}

class BST{
    constructor(){
        this.root = null;
    }
    add(data){
        const node = this.root;
        if(node === null){
            this.root = new Node(data);
            return;
        } else {
            const searchTree = function(node){
                if(data < node.data){
                    if(node.left === null){
                        node.left = new Node(data);
                    } else{
                        searchTree(node.left);
                    }
                } else if(data > node.data){
                    if(node.right === null){
                        node.right = new Node(data);
                    } else{
                        searchTree(node.right);
                    }
                } else{
                    return null;
                }
            }
            return searchTree(node);
        }
    }
    // In a binary search tree minimum node is at the leftest leaf node
    getMin(){
        const current = this.root;
        while(current.left !== null){
            current = current.left;
        }
        return current.val;
    }
    // In a BST maximum element is at the righmost corner
    getMax(){
        const current = this.root;
        while(current.right !== null){
            current = current.right;
        }
        return current.val;
    }
    find(data){
        const node = this.root;
        //Implementing DFS to find the node
        let stack = [node];
        while(stack.length>0){
            const current = stack.pop();
            if(current.val === data){
                return true;
            }
            if(current.right !== null){
                stack.push(current.right);
            } 
            if(current.left !== null){
                stack.push(current.left);
            }
        }
        return false;
    }
    remove(data) {
        const removeNode = function(node,data){
            if(node === null){
                return null;
            }
            if(data == node.data){
                //node has no children
                if(node.left === null && node.right === null) return null;
                //node has no left child
                if(node.left === null) return node.right;
                //node has no right child
                if(node.right === null) return node.left;
                //node has 2 children
                let tempnode = node.right;
                while(tempnode.left !== null){
                    tempnode= tempnode.left;
                } 
                node.data = tempnode.data;
                node.right = removeNode(node.right,tempnode.data);
                return node;
            } else if(data < node.data){
                node.left = removeNode(node.left,data);
            } else{
                node.right = removeNode(node.right,data);
            }
        }
        this.root = removeNode(this.root,data);
    }
}

File: trees/neetcode/binary-search-tree.py
class Node:
    def __init__(self,value) -> None:
        self.value = value
        self.left = None
        self.right = None
    
class binarySearchTree:
    def __init__(self):
        self.root = None
    
    def insert(self, value):
        if self.root is None:
            self.root = Node(value)
        else:
            def searchNInsert(node):
                if value < node.value:
                    if node.left == None:
                        node.left = Node(value)
                    else:
                        searchNInsert(node.left)
                elif value > node.value:
                    if node.right == None:
                        node.right = Node(value)
                    else:
                        searchNInsert(node.right)
                else:
                    return None
            searchNInsert(self.root)     
    def traverse(self):
        node = self.root
        def dfs(node):
            if node == None:
                return
            if node.left: dfs(node.left)
            if node.right: dfs(node.right)
            print(node.value)
        dfs(node)
    
    def isBalanced(self):
        node = self.root
        def dfs(node):
            if node is None: return [True,0]
            left, right = dfs(node.left), dfs(node.right)
            balanced = ((left[0]) and (right[0]) and abs(left[1]-right[1]) <=1)
            return [balanced,1+max(left[1],right[1])]
        return dfs(node)
        
    def remove(self,value):
        node = self.root
        if node is None:
            return None
        def removeNode(value,node):
            if node is None:
                return None
            if value == node.data:
                if node.left is None and node.right is None:
                    return None
                
                if node.left is None:
                    return node.right
                if node.right is None:
                    return node.left
                
                temp = node.right
                while(temp.left is not None):
                    temp = temp.left
                node.value = temp.value
                node.right = removeNode(temp.value,node.right)
                return node
            elif value < node.value:
                node.left = removeNode(value, node.left)
            else:
                node.right = removeNode(value,node.right)
        removeNode(value,node)

        
arr = [3,9,20,15,7]
bst = binarySearchTree()
for i in arr:
    bst.insert(i)

print(bst.isBalanced())

